src/basic/extract-word.c

```diff
@@ -29,54 +29,51 @@
 int extract_first_word(const char **p, char **ret, const char *separators, ExtractFlags flags) {
         _cleanup_free_ char *s = NULL;
         size_t allocated = 0, sz = 0;
+        char c;
         int r;
 
         char quote = 0;                 /* 0 or ' or " */
         bool backslash = false;         /* whether we've just seen a backslash */
-        bool separator = false;         /* whether we've just seen a separator */
-        bool start = true;              /* false means we're looking at a value */
 
         assert(p);
         assert(ret);
 
-        if (!separators)
-                separators = WHITESPACE;
-
         /* Bail early if called after last value or with no input */
         if (!*p)
                 goto finish_force_terminate;
+        c = **p;
+
+        if (!separators)
+                separators = WHITESPACE;
 
         /* Parses the first word of a string, and returns it in
          * *ret. Removes all quotes in the process. When parsing fails
          * (because of an uneven number of quotes or similar), leaves
          * the pointer *p at the first invalid character. */
 
-        for (;;) {
-                char c = **p;
-
-                if (start) {
-                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
-                                if (!GREEDY_REALLOC(s, allocated, sz+1))
-                                        return -ENOMEM;
+        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
+                if (!GREEDY_REALLOC(s, allocated, sz+1))
+                        return -ENOMEM;
 
-                        if (c == 0)
-                                goto finish_force_terminate;
-                        else if (strchr(separators, c)) {
+        for (;; (*p) ++, c = **p) {
+                if (c == 0)
+                        goto finish_force_terminate;
+                else if (strchr(separators, c)) {
+                        if (flags & EXTRACT_DONT_COALESCE_SEPARATORS) {
                                 (*p) ++;
-                                if (flags & EXTRACT_DONT_COALESCE_SEPARATORS)
-                                        goto finish_force_next;
-                                continue;
+                                goto finish_force_next;
                         }
-
+                } else {
                         /* We found a non-blank character, so we will always
                          * want to return a string (even if it is empty),
                          * allocate it here. */
                         if (!GREEDY_REALLOC(s, allocated, sz+1))
                                 return -ENOMEM;
-
-                        start = false;
+                        break;
                 }
+        }
 
+        for (;; (*p) ++, c = **p) {
```

Comment:
Agreed (in general), but if you look at tge code, you'll see it's
pre-existing, this is simply a refactor...

In this case we want it to update the pointer, because in some cases we
might want to ignore errors and resume where we stopped.

Cheers,
Filipe

On Fri, Nov 6, 2015, 11:37 Evgeny Vereshchagin <notifications@github.com>
wrote:

> In src/basic/extract-word.c
> <https://github.com/systemd/systemd/pull/1793#discussion_r44123653>:
>
> >
> > +        for (;; (*p) ++, c = **p) {
>
> should we change *p here?
>
> I like this rule
> <https://github.com/systemd/systemd/blob/master/CODING_STYLE#L198>:
>
>
>    - Do not write functions that clobber call-by-reference variables on
>    failure. Use temporary variables for these cases and change the passed in
>    variables only on success.
>
> See also: #1761 (comment)
> <https://github.com/systemd/systemd/pull/1761#discussion_r43815248>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/systemd/systemd/pull/1793/files#r44123653>.
>